***

# Machine Learning & Statistics Assessment - Scikit-Learn

***

Machine learning uses a variety of algorithms that continuously learn from data.  By using training data, we can build increasingly more precise models based on that data.They do this so they can characterise data, improve their output and predict outcomes. 
The power of machine learning is that we can constantly learn from data and help predict the future. “This powerful set of algorithms and models are being used across industries to improve processes and gain insights into patterns and anomalies within data” [2]
<br></br>
The open source Python package, scikit-learn enables us to implement these algorithms. This aim of this notebook is to provide a clear and concise overview of the Python package Scikit-Learn. This will be achieved by reseaching the package and demonstrating it's functionality through three algorithms of choice. 

[Link to Scikit-Learn Documentation](https://scikit-learn.org/stable/tutorial/index.html)

### What is Scikit Learn?

Scikit-learn is a library in Python that provides many unsupervised and supervised learning algorithms. It’s built upon some of the technology you might already be familiar with, like NumPy, pandas, and Matplotlib [1] Scikit-learn is a Python library created by David Cournapeau in 2007 as a Google Summer of Code project. Scikit-learn efficiently implements many of the machine learning algorithms. [3] The library is focused on modeling data. Scikit-learn aims on bringing machine learning to non-specialists. It puts an importance on its ease of use, performance, documentation, and API consistency. It has minimal dependencies and is distributed under the simplified BSD license, encouraging its use in both academic and commercial settings [4] Scikit-learn is built upon NumPy, SciPy and Matplotlib.

#### Set Up

In [24]:
# Importing Libaries
import sklearn.datasets as sk

import numpy as np

import pandas as pd

import matplotlib.pyplot as plt

import seaborn as sns

import sklearn.model_selection as mod

### Supervised Learning

A supervised machine learning algorithm (as opposed to an unsupervised machine learning algorithm) is one that relies on labeled input data to learn a function that produces an appropriate output when given new unlabeled data.
Supervised machine learning algorithms are used to solve classification or regression problems.
<br></br>

Some of the most commonly used supervised learning algorithms are:

- Linear Regression
- Logistic Regression 
- K Nearest Neighbour
- Neural Networks
- Naive Bayers
- Random Forest
- Support Vector Machines

Supervised machine learning is used to solve classification or regression problem.

A *classification problem* has a discrete value as its output. i.e. There is no middle ground, it’s a yes or no, true or false.

A *regression problem* has a real number (float) as its output.  


### Unsupervised Learning

An unsupervised machine learning algorithm  tries to learn the basic structure of the data to give more insight into the data.Unlike supervised learning which tries to learn a function that will allow us to make predictions given some new unlabeled data, unsupervised learning makes use of input data without any labels. s. From this data, the algorithm discovers patterns which help solve clustering or association problems. This is particularly useful when subject matter experts are unsure of common properties within a data set.
<br></br>

Unsupervised machine learning, mainly deals with finding a pattern in a collection of uncategorized data. Clustering algorithms will process the data and find natural clusters(groups) if they exist. 

### K Nearest Neighbour

[K Nearest Neighbour documentation](https://scikit-learn.org/stable/modules/neighbors.html)

K Nearest Neighbour (KNN) works on the principle assuming every data point falling in near to each other is falling in the same class. In other words, it classifies a new data point based on similarity. KNN is a simple, easy-to-implement supervised machine learning algorithm that can be used to solve both classification and regression problems.
The KNN algorithm assumes that similar things exist in close proximity. In other words, similar things are near to each other. KNN is typically used for recommendation engines and image recognition.

Going to use the Breast Cancer Wisconsin (diagnostic) dataset — use ML to diagnose cancer scans as benign (does not spread to the rest of the body) or malignant (spreads to rest of the body)


#### Setup

In [2]:
# Importing Libaries

from sklearn.model_selection import train_test_split

from sklearn.neighbors import KNeighborsClassifier

import sklearn.neighbors as knn

#### Data

In [13]:
# Loading the data
from sklearn.datasets import load_breast_cancer
cancer = load_breast_cancer()

In [14]:
print(cancer.DESCR) 

.. _breast_cancer_dataset:

Breast cancer wisconsin (diagnostic) dataset
--------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 569

    :Number of Attributes: 30 numeric, predictive attributes and the class

    :Attribute Information:
        - radius (mean of distances from center to points on the perimeter)
        - texture (standard deviation of gray-scale values)
        - perimeter
        - area
        - smoothness (local variation in radius lengths)
        - compactness (perimeter^2 / area - 1.0)
        - concavity (severity of concave portions of the contour)
        - concave points (number of concave portions of the contour)
        - symmetry
        - fractal dimension ("coastline approximation" - 1)

        The mean, standard error, and "worst" or largest (mean of the three
        worst/largest values) of these features were computed for each image,
        resulting in 30 features.  For instance, field 0 is Mean Radi

In [11]:
type(cancer)

sklearn.utils.Bunch

These load functions (such as load_breast_cancer()) don’t return data in the tabular format we may expect. They return a Bunch object. A bunch is a container object for datasets. It's like a dictionary object that stores data as keys and values. 

https://stackoverflow.com/questions/38105539/how-to-convert-a-scikit-learn-dataset-to-a-pandas-dataset/46379878#46379878

In [15]:
print(cancer.keys())

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename'])


<br></br>
data is all features of the data (attributes that help identify if the tumor is malignant or benign) 
Target is the target data (the variable you want to predict i.e. if the tumor is malignant or benign) 

The remaining keys serve a descriptive purpose. It’s important to note that all of Scikit-Learn datasets are divided into data and target. data represents the features, which are the variables that help the model learn how to predict. target includes the actual labels. In our case, the target data is one column classifies the tumor as either 0 indicating malignant or 1 for benign.
https://towardsdatascience.com/how-to-use-scikit-learn-datasets-for-machine-learning-d6493b38eca3

In [16]:
# Read the DataFrame, first using the feature data
df = pd.DataFrame(cancer.data, columns=cancer.feature_names)

In [17]:
# Add a target column, and fill it with the target data
df['target'] = cancer.target

In [19]:
# Having a look at the data
df

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0          17.99         10.38          122.80     1001.0          0.11840   
1          20.57         17.77          132.90     1326.0          0.08474   
2          19.69         21.25          130.00     1203.0          0.10960   
3          11.42         20.38           77.58      386.1          0.14250   
4          20.29         14.34          135.10     1297.0          0.10030   
..           ...           ...             ...        ...              ...   
564        21.56         22.39          142.00     1479.0          0.11100   
565        20.13         28.25          131.20     1261.0          0.09780   
566        16.60         28.08          108.30      858.1          0.08455   
567        20.60         29.33          140.10     1265.0          0.11780   
568         7.76         24.54           47.92      181.0          0.05263   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
0             0.27760         0.30010              0.14710         0.2419   
1             0.07864         0.08690              0.07017         0.1812   
2             0.15990         0.19740              0.12790         0.2069   
3             0.28390         0.24140              0.10520         0.2597   
4             0.13280         0.19800              0.10430         0.1809   
..                ...             ...                  ...            ...   
564           0.11590         0.24390              0.13890         0.1726   
565           0.10340         0.14400              0.09791         0.1752   
566           0.10230         0.09251              0.05302         0.1590   
567           0.27700         0.35140              0.15200         0.2397   
568           0.04362         0.00000              0.00000         0.1587   

     mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0                   0.07871  ...          17.33           184.60      2019.0   
1                   0.05667  ...          23.41           158.80      1956.0   
2                   0.05999  ...          25.53           152.50      1709.0   
3                   0.09744  ...          26.50            98.87       567.7   
4                   0.05883  ...          16.67           152.20      1575.0   
..                      ...  ...            ...              ...         ...   
564                 0.05623  ...          26.40           166.10      2027.0   
565                 0.05533  ...          38.25           155.00      1731.0   
566                 0.05648  ...          34.12           126.70      1124.0   
567                 0.07016  ...          39.42           184.60      1821.0   
568                 0.05884  ...          30.37            59.16       268.6   

     worst smoothness  worst compactness  worst concavity  \
0             0.16220            0.66560           0.7119   
1             0.12380            0.18660           0.2416   
2             0.14440            0.42450           0.4504   
3             0.20980            0.86630           0.6869   
4             0.13740            0.20500           0.4000   
..                ...                ...              ...   
564           0.14100            0.21130           0.4107   
565           0.11660            0.19220           0.3215   
566           0.11390            0.30940           0.3403   
567           0.16500            0.86810           0.9387   
568           0.08996            0.06444           0.0000   

     worst concave points  worst symmetry  worst fractal dimension  target  
0                  0.2654          0.4601                  0.11890       0  
1                  0.1860          0.2750                  0.08902       0  
2                  0.2430          0.3613                  0.08758       0  
3                  0.2575          0.6638                  0.17300       0  
4                  0.1625          0.2364                  0.076

In [21]:
# Summary
df.describe()

mean radius  mean texture  mean perimeter    mean area  \
count   569.000000    569.000000      569.000000   569.000000   
mean     14.127292     19.289649       91.969033   654.889104   
std       3.524049      4.301036       24.298981   351.914129   
min       6.981000      9.710000       43.790000   143.500000   
25%      11.700000     16.170000       75.170000   420.300000   
50%      13.370000     18.840000       86.240000   551.100000   
75%      15.780000     21.800000      104.100000   782.700000   
max      28.110000     39.280000      188.500000  2501.000000   

       mean smoothness  mean compactness  mean concavity  mean concave points  \
count       569.000000        569.000000      569.000000           569.000000   
mean          0.096360          0.104341        0.088799             0.048919   
std           0.014064          0.052813        0.079720             0.038803   
min           0.052630          0.019380        0.000000             0.000000   
25%           0.086370          0.064920        0.029560             0.020310   
50%           0.095870          0.092630        0.061540             0.033500   
75%           0.105300          0.130400        0.130700             0.074000   
max           0.163400          0.345400        0.426800             0.201200   

       mean symmetry  mean fractal dimension  ...  worst texture  \
count     569.000000              569.000000  ...     569.000000   
mean        0.181162                0.062798  ...      25.677223   
std         0.027414                0.007060  ...       6.146258   
min         0.106000                0.049960  ...      12.020000   
25%         0.161900                0.057700  ...      21.080000   
50%         0.179200                0.061540  ...      25.410000   
75%         0.195700                0.066120  ...      29.720000   
max         0.304000                0.097440  ...      49.540000   

       worst perimeter   worst area  worst smoothness  worst compactness  \
count       569.000000   569.000000        569.000000         569.000000   
mean        107.261213   880.583128          0.132369           0.254265   
std          33.602542   569.356993          0.022832           0.157336   
min          50.410000   185.200000          0.071170           0.027290   
25%          84.110000   515.300000          0.116600           0.147200   
50%          97.660000   686.500000          0.131300           0.211900   
75%         125.400000  1084.000000          0.146000           0.339100   
max         251.200000  4254.000000          0.222600           1.058000   

       worst concavity  worst concave points  worst symmetry  \
count       569.000000            569.000000      569.000000   
mean          0.272188              0.114606        0.290076   
std           0.208624              0.065732        0.061867   
min           0.000000              0.000000        0.156500   
25%           0.114500              0.064930        0.250400   
50%           0.226700              0.099930        0.282200   
75%           0.382900              0.161400        0.317900   
max           1.252000              0.291000        0.663800   

       worst fractal dimension      target  
count               569.000000  569.000000  
mean                  0.083946    0.627417  
std                   0.018061    0.483918  
min                   0.055040    0.000000  
25%                   0.071460    0.000000  
50%                   0.080040    1.000000  
75%                   0.092080    1.000000  
max                   0.207500    1.000000  

[8 rows x 31 columns]

In [22]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 31 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   mean radius              569 non-null    float64
 1   mean texture             569 non-null    float64
 2   mean perimeter           569 non-null    float64
 3   mean area                569 non-null    float64
 4   mean smoothness          569 non-null    float64
 5   mean compactness         569 non-null    float64
 6   mean concavity           569 non-null    float64
 7   mean concave points      569 non-null    float64
 8   mean symmetry            569 non-null    float64
 9   mean fractal dimension   569 non-null    float64
 10  radius error             569 non-null    float64
 11  texture error            569 non-null    float64
 12  perimeter error          569 non-null    float64
 13  area error               569 non-null    float64
 14  smoothness error         5

- This dataframe consists of 569 rows x 31 columns. 
- There are no missing (Null/NaN) values. 
- All vaules are a float type.

## References

[1] https://www.codecademy.com/articles/scikit-learn
[2]https://www.ibm.com/downloads/cas/GB8ZMQZ3#:~:text=Machine%20learning%20is%20a%20form,describe%20data%2C%20and%20predict%20outcomes.
[3] https://books.google.ie/books?id=HnetDwAAQBAJ&printsec=frontcover&dq=what+is+scikit+learn&hl=en&sa=X&ved=2ahUKEwjK0I3TzcXzAhXBoVwKHYM4CHEQ6AF6BAgKEAI#v=onepage&q=what%20is%20scikit%20learn&f=false
[4] https://jmlr.org/papers/v12/pedregosa11a.html


## End